### PR TITLE
New filter 'pdf2text' to extract text from PDF

### PR DESCRIPTION
**Pull request by mborsetti**
A filter 'pdf2text' is created to extract text from a PDF file; this allows tracking of changes to the contents of PDF files.

It is based on the pdftotext library (https://github.com/jalan/pdftotext), which in turn is based on poppler (https://poppler.freedesktop.org/).  README has been updated with the additional requirements.

It only works on bytes, so I also had to modify code in the job.py library: if this filter is specified, a bytes object is returned instead of a str.  Longer term this probably should be a more generic option.

### CODE REVIEW

**Comment by cfbao** [comment]
You've merged your master branch (which contains all changes you want to make) into this single feature branch, which resulted in this PR containing all your proposed changes (instead of a single one).

Your commit [d46b4c9](https://github.com/mborsetti/urlwatch/tree/d46b4c90cccb3911e508e93ec8015335eb379d57) seems to be a good place for this branch and PR.
You can run the following git commands to set this branch to this commit:
```
git checkout pdf2text
git reset --hard d46b4c9
git push --force
```

**Comment by cfbao** [comment]
My first thought on jobs/filters' input/output types (str/bytes) handling:

A filter class can declare its supported input and output types as class variables
```python
class ExampleFilter(FilterBase):
    INPUT_TYPE = str
    OUTPUT_TYPE = [str, bytes]
```
The `Filter.filter` method and `job.retrieve` method should have an optional argument `output_type` that's used when multiple output types are supported.

The logic of matching and using correct input/output types should be placed outside of jobs or filters, but inside `JobState.process`.

The base filter class should declare default input/output types both as `str` for backwards compatibility, to not break users' custom hooks.

**Comment by mborsetti** [comment]
> You've merged your master branch (which contains all changes you want to make) into this single feature branch, which resulted in this PR containing all your proposed changes (instead of a single one).
> 
> Your commit [d46b4c9](https://github.com/mborsetti/urlwatch/tree/d46b4c90cccb3911e508e93ec8015335eb379d57) seems to be a good place for this branch and PR.
> You can run the following git commands to set this branch to this commit:
> 
> ```
> git checkout pdf2text
> git reset --hard d46b4c9
> git push --force
> ```

Thanks, and apologies for the inadvertent mess (I still have to understand how I did it, I think it was on the website).

File: lib/urlwatch/jobs.py [hunk 1]
```diff
@@ -284,6 +287,11 @@ def retrieve(self, job_state):
         # Save ETag from response into job_state, which will be saved in cache
         job_state.etag = response.headers.get('ETag')
 
+        # If 'pdf2text' filter is selected, return bytes content as opposed to str
+        # as it's required by the library used by that filter
+        if 'pdf2text' in self.filter or any('pdf2text' in subfilter for subfilter in self.filter):
```
Comment (thp) [inline lib/urlwatch/jobs.py]: Can you turn this expression (`'pdf2text' in self.filter or any('pdf2text' in subfilter for subfilter in self.filter)`) into a helper function like `self.needs_binary_data()`? Or maybe we could make it so that this job has a "binary: true" flag or something? IMHO it would be better to do it this way (or make it so that the filter object has a "needs_binary = True" or something and find it out this way -- right now it's a bit hard to maintain).

**Comment by thp** [comment]
@mborsetti Did you have time to look at the PR review comments already?

**Comment by mborsetti** [comment]
@thp,

On my todo list now; should be done tonight UTC +08.

**Comment by mborsetti** [comment]
I am sorry, but due to some developments relating to COVID-19 I have not
been able to do any and highly doubt I will be able to contribute at all in
the near term.

**Comment by mborsetti** [comment]
Hi Tom,
Had a bit of time to look at this. I am still a bit confused about the overall architecture. If a switch indicating "filter needs binary data" or "job is binary data" were to be created, where would it be set (i.e. where would you do the conditional `if 'pdf2text' in self.filter or any('pdf2text' in subfilter for subfilter in self.filter):`?

File: lib/urlwatch/filters.py [hunk 2]
```diff
@@ -198,6 +198,29 @@ def filter(self, data, subfilter=None):
...
+
+    def filter(self, data, subfilter=None):
+        # data must be bytes
+
+        if subfilter is None:
+            password = ''
+        elif isinstance(subfilter, dict):
+            password = subfilter['password']
+        elif isinstance(subfilter, str):
+            password = subfilter
```
Comment (thp) [inline lib/urlwatch/filters.py:218]: IMHO it's totally fine to require that the subfilter must be specified as dict, so this can be removed.

File: docs/source/dependencies.rst [hunk 3]
```diff
@@ -38,7 +38,7 @@ Where ``<packagename>`` is one of the following:
 +-------------------------+---------------------------------------------------------------------+
 | Feature                 | Python package(s) to install                                        |
 +=========================+=====================================================================+
-| Pushover reporter       | `chump <https://github.com/karanlyons/chump/>`__                    |
+| Pushover reporter       | `chump <https://github.com/karanlyons/chump>`__                     |
```
Comment (thp) [inline docs/source/dependencies.rst:41]: Unrelated change to this PR, can you move it into a separate PR (yeah, it's a one-character change, but just to keep it separate).

Comment (mborsetti) [reply docs/source/dependencies.rst:41]: Fair enough!

File: lib/urlwatch/filters.py (hunk 2 above)
Comment (mborsetti) [reply lib/urlwatch/filters.py:218]: 'k

File: lib/urlwatch/jobs.py [hunk 4]
```diff
@@ -284,6 +284,11 @@ def retrieve(self, job_state):
         # Save ETag from response into job_state, which will be saved in cache
         job_state.etag = response.headers.get('ETag')
 
+        # If 'pdf2text' filter is selected, return content in bytes instead of in unicode
+        # as that's what's required by the library used by that filter
+        if self.filter and ('pdf2text' in self.filter or any('pdf2text' in subfilter for subfilter in self.filter)):
+            return response.content
+
```
Comment (thp) [inline lib/urlwatch/jobs.py:291]: "subfilter" is what we call the "parameters" of the filter. Here's my proposal:

- Add `__uses_bytes__ = True` to `Pdf2TextFilter`

Then, to get all filter names:

    bytes_filters = [name for name, class_ in FilterBase.__subclasses__.items()
                     if getattr(class_, '__uses_bytes__', False)]

(by default this should evaluate to `['pdf2text']` if `__uses__bytes__ = True` is set on the `Pdf2TextFilter` class)

Then you can do:

    for filter_name in bytes_filters:
        if self.filter and (filter_name in self.filter or any(filter_name in filter for filter in self.filter)):
            # a bytes filter is selected, return content in bytes instead of in unicode
            # as that's what's required by the library used by that filter
            return response.content

Of course, it's tricky if you try to chain filters... This assumes that the `pdf2text` filter is the "first" filter. The following filter chain won't work (properly):

```
    filter:
        - html2text
        - pdf2text
```

So maybe even check if it's the only filter or if `isinstance(self.filter, list)` then only check if the first filter is a bytes filter (and maybe also check that all others filters are NOT bytes filters).

Comment (mborsetti) [reply lib/urlwatch/jobs.py:291]: Checking the first filter makes more sense since Job only creates data to the first filter

The issue with chaining only makes sense if you have filters that output bytes, which you don't have.  Also, I can't think what would be a use case for one of them.  Will document that pdf2text needs to be the *first* filter and work on throwing an error if it's fed a str type, which would happen if it's not the first filter.

**Comment by thp** [comment]
Thanks, will merge once CI is happy.

File: lib/urlwatch/jobs.py [hunk 5]
```diff
@@ -198,6 +199,13 @@ def retrieve(self, job_state):
         if result != 0:
             raise ShellError(result)
 
+    bytes_filters = [name for name, class_ in FilterBase.__subclasses__.items()
+                     if getattr(class_, '__uses_bytes__', False)]
```
Comment (thp) [inline lib/urlwatch/jobs.py:203]: Now that is in two places, maybe we can make this a classmethod on `FilterBase`?

    @classmethod
    def is_bytes_filter(cls, filter):
        return (filter in [name for name, class_ in FilterBase.__subclasses__.items()
                           if getattr(class_, '__uses_bytes__', False)])

Comment (mborsetti) [reply lib/urlwatch/jobs.py]: Oh, I am learning lots from you.  On to it.

File: lib/urlwatch/jobs.py [hunk 6]
```diff
@@ -198,6 +199,11 @@ def retrieve(self, job_state):
         if result != 0:
             raise ShellError(result)
 
+        # If the first filter is a bytes filter, return content in bytes instead of in unicode
+        # as that's what's required by the library used by that filter
+        if self.filter and (FilterBase(self, self.filter) or FilterBase(self, self.filter[0])):
```
Comment (thp) [inline lib/urlwatch/jobs.py:204]:
```python
if self.filter and (Filterbase.is_bytes_filter(self.filter) or FilterBase.is_bytes_filter(self.filter[0]):
```

Comment (mborsetti) [reply lib/urlwatch/jobs.py]: Yeah, my internal test needs upgrading; thanks for the hint.

The correct code is actually `if self.filter and (FilterBase.is_bytes_filter(self.filter)
                            or any(FilterBase.is_bytes_filter(filter) for filter in self.filter[0])):`, although it's still technically incorrect because I should be checking the first key only, not all (using `any()`)

Comment (mborsetti) [reply lib/urlwatch/jobs.py]: Now only checking first key.

I can't think of anything else.  Thanks for your guidance and patience.

**Comment by thp** [comment]
Thanks for working on this! Merged :)
